### PR TITLE
8293345: SunPKCS11 provider checks on PKCS11 Mechanism are problematic

### PR DESCRIPTION
**Pull request by valeriepeng**
Existing legacy mechanism check disables mechanism(s) when the support is partial, e.g. supports decryption but not encryption, or supports verification but not signing. Some mechanisms can be used for both encryption/decryption and sign/verify such as RSA related ones. If the particular mechanism supports sign/verify/decryption but not encryption, it'd be disabled as a result. Fine tune the legacy mechanism check with the service type, i.e. supports encryption for Cipher,  sign for Signature, so the mechanism is disabled based on the service type.
For completeness sake, I also added a PKCS11 provider configuration option to control this.  If not set, SunPKCS11 provider will disable legacy mechanisms by default.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8329300](https://bugs.openjdk.org/browse/JDK-8329300) to be approved

### Issues
 * [JDK-8293345](https://bugs.openjdk.org/browse/JDK-8293345): SunPKCS11 provider checks on PKCS11 Mechanism are problematic (**Enhancement** - P3)
 * [JDK-8329300](https://bugs.openjdk.org/browse/JDK-8329300): SunPKCS11 provider checks on PKCS11 Mechanism are problematic (**CSR**)

### Reviewers
 * [Daniel Jeliński](https://openjdk.org/census#djelinski) (@djelinski - **Reviewer**)
 * [Weijun Wang](https://openjdk.org/census#weijun) (@wangweij - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18387/head:pull/18387` \
`$ git checkout pull/18387`

Update a local copy of the PR: \
`$ git checkout pull/18387` \
`$ git pull https://git.openjdk.org/jdk.git pull/18387/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18387`

View PR using the GUI difftool: \
`$ git pr show -t 18387`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18387.diff">https://git.openjdk.org/jdk/pull/18387.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18387#issuecomment-2008586764)

### CODE REVIEW

**Review by djelinski — APPROVED** [review]
LGTM

**Comment by seanjmullan** [comment]
Is "disableLegacy" a standard PKCS11 attribute we are introducing support for? If so, I think a CSR is probably needed as it is kind of like a system property.

**Comment by valeriepeng** [comment]
> Is "disableLegacy" a standard PKCS11 attribute we are introducing support for? If so, I think a CSR is probably needed as it is kind of like a system property.

It's a configuration file option. Just for the uncommon case if users would like to disable the check entirely. I can file a CSR on this. Thanks for the comment.

**Comment by valeriepeng** [comment]
> LGTM

Thanks for the review~

**Review by djelinski — APPROVED** [review]

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/SunPKCS11.java
Comment (mcpowers) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/SunPKCS11.java:1344]: I'm curious why this check was moved down in the file. The only advantage I see is that you have access to `d.type`.

**Review by mcpowers — COMMENTED** [review]
What about testing?

**Comment by valeriepeng** [comment]
> What about testing?

I tested the change manually by changing some behavior to simulate the case. However, this cannot be done by regression test since NSS does not have legacy mechanisms.

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/SunPKCS11.java
Comment (valeriepeng) [reply src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/SunPKCS11.java:1344]: Yes, this is the purpose, e.g. perform the check based on the type of registered service. The earlier check does not take into account of the registered service and disables a mechanism completely even when it can still be used for some.
For example, a mechanism supports decryption, signing, and verification will be disabled completely with the earlier check, but will still be enabled for Signature with this proposed change.

**Comment by dems54** [comment]
@valeriepeng can this feature be backported to Java 17 ?

**Comment by valeriepeng** [comment]
> @valeriepeng can this feature be backported to Java 17 ?

I suppose so, the PKCS11 configuration attribute part which requires CSR can be separated out if needed.

**Comment by valeriepeng** [comment]
@mcpowers I am about to leave for vacation. Will wait for your review and resume on this PR after I return. Thanks!

**Review by wangweij — APPROVED** [review]

**Comment by mcpowers** [comment]
Your changes look fine to me.

**Comment by dems54** [comment]
Many thanks @valeriepeng for this feature.
Possible to backport on JDK 17 or 21 ?

**Comment by dems54** [comment]
UP @valeriepeng possible to backport PKCS11 configuration attribute part  on JDK 17 or 21 ?

**Comment by valeriepeng** [comment]
> UP @valeriepeng possible to backport PKCS11 configuration attribute part on JDK 17 or 21 ?

I am not sure, let me check and get back to you.

**Comment by valeriepeng** [comment]
> UP @valeriepeng possible to backport PKCS11 configuration attribute part on JDK 17 or 21 ?

Based  on the feedback that I got, it should be ok to backport the PKCS11 configuration attribute, separate CSRs would be needed for that though.

**Comment by dems54** [comment]
Many thanks @valeriepeng !
When could the backport to JDK17 and 21 be done ?
We are very excited to be able to update our clients with a maintained java version

**Comment by valeriepeng** [comment]
> Many thanks @valeriepeng ! When could the backport to JDK17 and 21 be done ? We are very excited to be able to update our clients with a maintained java version

This would be subject to the 21u and 17u Maintainer. I've opened the backport request against 17-pool and 21-pool but someone needs to file the CSR and do the actual backport.

**Pull request closed without merging**
